File: src/models/Document.ts
import mongoose, { Schema, Document as MongooseDocument } from 'mongoose';

export interface IDocument extends MongooseDocument {
  uploaderId: mongoose.Types.ObjectId;
  title: string;
  originalFilename: string;
  url: string;
  publicId: string; // Cloudinary public_id
  documentType: string;
  status: 'Pending' | 'Processing' | 'Validation_Pending' | 'Completed' | 'Failed';
  confidenceScore?: number;
  extractedData?: Record<string, any>;
  validationErrors?: Record<string, string>;
  createdAt: Date;
  updatedAt: Date;
}

const DocumentSchema: Schema = new Schema(
  {
    uploaderId: { type: Schema.Types.ObjectId, ref: 'User', required: true, index: true },
    title: { type: String, required: true },
    originalFilename: { type: String, required: true },
    url: { type: String, required: true },
    publicId: { type: String, required: true },
    documentType: {
      type: String,
      default: 'Unknown',
      // We will allow open strings initially but recommend: Invoice, BankStatement, ITR, SalarySlip, etc.
    },
    status: {
      type: String,
      enum: ['Pending', 'Processing', 'Validation_Pending', 'Completed', 'Failed'],
      default: 'Pending',
      index: true,
    },
    confidenceScore: { type: Number },
    extractedData: { type: Schema.Types.Mixed },
    validationErrors: { type: Schema.Types.Mixed },
  },
  {
    timestamps: true,
  }
);

export const Document = mongoose.models.Document || mongoose.model<IDocument>('Document', DocumentSchema);

File: backend/models/document.py
from sqlalchemy import Column, Integer, String, Float, DateTime, Enum, ForeignKey, JSON
from sqlalchemy.orm import relationship
from datetime import datetime, timezone
import enum
from backend.core.database import Base
from backend.models.user import generate_uuid

class DocumentStatusEnum(str, enum.Enum):
    Pending = 'Pending'
    Processing = 'Processing'
    Validation_Pending = 'Validation_Pending'
    Completed = 'Completed'
    Failed = 'Failed'

class Document(Base):
    __tablename__ = "documents"

    id = Column(String, primary_key=True, default=generate_uuid, index=True)
    uploader_id = Column(String, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    title = Column(String, nullable=False)
    original_filename = Column(String, nullable=False)
    url = Column(String, nullable=False)
    public_id = Column(String, nullable=False)
    document_type = Column(String, default="Unknown")
    status = Column(Enum(DocumentStatusEnum), default=DocumentStatusEnum.Pending, index=True)
    confidence_score = Column(Float, nullable=True)
    
    # Store dynamic parsed data and errors
    extracted_data = Column(JSON, nullable=True)
    validation_errors = Column(JSON, nullable=True)
    
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    uploader = relationship("User", back_populates="documents")
    audit_logs = relationship("AuditLog", back_populates="document")
